nomad/structs/structs.go

```diff
@@ -8386,17 +8385,16 @@ func (a *Affinity) Copy() *Affinity {
 	if a == nil {
 		return nil
 	}
-	na := new(Affinity)
-	*na = *a
-	return na
+	return &Affinity{
+		LTarget: a.LTarget,
+		RTarget: a.RTarget,
+		Operand: a.Operand,
+		Weight:  a.Weight,
+	}
```

Comment:
I'm conflicted about this transition in this PR and the constraints PR. The previous way copes with changes to the structs better and is more common approach in this file; but I like the explicitness here. Would like to know how you view it and whether we should change the rest of Copy functions?

Reply:
I think it comes down to which kind of bug(s) do you want to track down in the future. The explicit assignment here implies you'll have a zero-value or a nil-value if you forget to add the new field. 

OTOH using the de-reference copy trick creates a shallow copy - which if you add a pointer field and forget to deep copy it, it will now be shared and mutable by either owner.

I my experience NPE's and zero-values have been easier to track down than shared mutable state, which is why I kind of prefer these explicit copies. 

Maybe the best thing would actually be to use something like https://pkg.go.dev/github.com/getlantern/deepcopy, or have Go add copy constructors to the language